draft-ietf-quic-tls.md

```diff
@@ -724,6 +689,37 @@ This results in abandoning loss recovery state for the Initial encryption level
 and ignoring any outstanding Initial packets.
 
 
+### Discarding Handshake Keys
+
+In this document, the TLS handshake is considered confirmed from an endpoint's
+perspective when both of the following two conditions are met: the handshake is
+complete from this endpoint’s perspective and the endpoint has received an
+ACK for a packet sent with 1-RTT keys.  This second condition can be
+implemented as checking whether largest_recvd_ack_1rtt >= first_1rtt_pn, where
+largest_recvd_ack_1rtt represents the highest value of the Largest Acknowledged
+field in any received 1-RTT ACK frame, and where first_1rtt_pn represents the
+lowest packet number that this endpoint has sent encrypted with 1-RTT keys
+(first_1rtt_pn is always 0 on the server since servers cannot send 0-RTT
+packets).
+
+An endpoint MUST NOT discard its handshake keys until the TLS handshake is
+confirmed from its perspective.  An endpoint MAY discard its handshake keys
+as soon as it has confirmed the handshake.  Endpoints SHOULD repeatedly send
```

Comment:
Why not: MUST discard as soon as it has confirmed the handshake?

Reply:
I don't think discarding handshake keys provides security benefits (unlike initial keys) so I don't see a reason to mandate discarding.

Reply:
Discarding keys is always advisable from a crypto hygiene perspective. I agree it's a lot less critical than discarding the Initial keys. Maybe we can make it a SHOULD?

Reply:
Agreed. Done.

Reply:
Sorry for re-opening the discussion here, but we now have a MUST discard for 0-RTT keys. All that an attacker can do with 0-RTT packets is to replay them, and the same applies to Handshake keys. Shouldn't we use a MUST for Handshake keys as well then?

Reply:
I don't think it's quite the same. The server discarding 0-RTT keys prevents a buggy client from sending more requests over 0-RTT when it has 1-RTT. A client can't send application data with handshake keys.